### PR TITLE
Add seeking to `Membrane.WAV.Serializer`

### PR DESCRIPTION
**Pull request by balins**
If `Membrane.File.SeekEvent` is available, `Membrane.WAV.Serializer` will emit event that updates invalid WAV header. Otherwise, `Membrane.WAV.Serializer` works like before.

### CODE REVIEW

File: lib/membrane_wav/serializer_bin/interceptor.ex [hunk 1]
```diff
@@ -0,0 +1,61 @@
+defmodule Membrane.WAV.SerializerBin.Interceptor do
```
Comment (bblaszkow06) [inline lib/membrane_wav/serializer_bin/interceptor.ex]: OK, seeing the implementation I think I misled you and using a bin wasn't a good idea, sorry about that. I was convinced the support for seeking in the file is needed at a parent of the serializer. While it obviously can be done inside the pipeline.

First of all, the default way to serialize to WAV should generate a valid file, this shouldn't be an optional behaviour, so I'd move the logic from Interceptor to Serializer. Then you can add a flag to turn it off - to allow using it with something different than file sink. The doc there should point to the Postprocessing module.

File: lib/membrane_wav/serializer_bin/interceptor.ex [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  @impl true
+  def handle_process(:input, buffer, _context, %{header_length: 0} = state) do
+    state = Map.put(state, :header_length, byte_size(buffer.payload))
+    {{:ok, buffer: {:output, buffer}, redemand: :output}, state}
+  end
+
+  def handle_process(:input, buffer, _context, %{data_length: data_length} = state) do
+    state = Map.put(state, :data_length, data_length + byte_size(buffer.payload))
+    {{:ok, buffer: {:output, buffer}, redemand: :output}, state}
+  end
```
Comment (bblaszkow06) [inline lib/membrane_wav/serializer_bin/interceptor.ex]: You can reduce the number of callback calls by using `handle_process_list` and either `Enum.reduce` or `Enum.map` + `Enum.sum`

The former function clause is harder to translate to `_list` but after merge with `Serializer` the need to intercept header should disappear

File: lib/membrane_wav/serializer_bin/interceptor.ex [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  def handle_end_of_stream(
+        :input,
+        _context,
+        %{header_length: header_length, data_length: data_length} = state
+      ) do
+    file_length = header_length + data_length
+
+    actions = [
+      event: {:output, %File.SeekEvent{position: @file_length_offset}},
+      buffer: {:output, %Buffer{payload: <<file_length - 8::32-little>>}},
```
Comment (bblaszkow06) [inline lib/membrane_wav/serializer_bin/interceptor.ex]: I'm guessing the `-8` comes from skipping `RIFF` and data_length - I think it will be possible to take it into account earlier after the merge, keep an eye for it.

Anyway, this feels a bit enigmatic so deserves either a comment or being assigned to a descriptive variable (you may apply that to the `Postprcessing` module as well)

File: lib/membrane_wav/serializer_bin.ex [hunk 4]
```diff
@@ -0,0 +1,56 @@
+defmodule Membrane.WAV.SerializerBin do
```
Comment (bblaszkow06) [inline lib/membrane_wav/serializer_bin.ex]: After the suggested changes this module won't be much of a help, especially since it adds a dependency on `file_plugin`. Of course, we could do some magic here and make `file_plugin` dependency optional (I had in mind a runtime check on `File.Sink` module existence, raising an error if it's missing and instructing to add `file_plugin` do deps, something similar to what is done for example in Tesla adapters)

TL;DR This can be removed, sorry

File: test/serializer_bin/serializer_bin_test.exs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+  @input_path Path.expand("../fixtures/input.wav", __DIR__)
+  @reference_path Path.expand("../fixtures/reference_processed.wav", __DIR__)
+
+  setup_all do
+    output_path = Path.join([System.tmp_dir!(), "output.wav"])
+    on_exit(fn -> File.rm!(output_path) end)
+
+    [output_path: output_path]
+  end
```
Comment (bblaszkow06) [inline test/serializer_bin/serializer_bin_test.exs]: It's a fairly recent discovery, but systematically introduced to tests among elements - you can use the `tmp_dir` tag of ExUnit https://hexdocs.pm/ex_unit/ExUnit.Case.html#module-tmp-dir

File: test/serializer_bin/serializer_bin_test.exs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+      ]
+
+      links = [link(:file_source) |> to(:parser) |> to(:serializer_bin)]
+
+      options = %Pipeline.Options{elements: elements, links: links}
+
+      {:ok, pid} = Pipeline.start_link(options)
+      :ok = Pipeline.play(pid)
+
+      assert_pipeline_notified(pid, :serializer_bin, :end_of_stream)
```
Comment (bblaszkow06) [inline test/serializer_bin/serializer_bin_test.exs]: I'd rather see [`assert_end_of_stream`](https://hexdocs.pm/membrane_core/Membrane.Testing.Assertions.html#assert_end_of_stream/4) here

**Review by bblaszkow06 — CHANGES_REQUESTED** [review]

File: test/serializer_bin/serializer_bin_test.exs (hunk 5 above)
Comment (balins) [reply test/serializer_bin/serializer_bin_test.exs]: Thanks, didn't know this one. Shame that the created `"tmp"` directory in project dir isn't automatically removed though.
